spectrocloud/resource_workspace.go

```diff
@@ -0,0 +1,304 @@
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"name": {
+							Type:     schema.TypeString,
+							Required: true,
+						},
+						"resource_allocation": {
+							Type:     schema.TypeMap,
+							Required: true,
+							Elem: &schema.Schema{
+								Type: schema.TypeString,
+							},
+						},
+						"images_blacklist": {
+							Type:     schema.TypeList,
+							Optional: true,
+							Elem: &schema.Schema{
+								Type: schema.TypeString,
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+func resourceWorkspaceCreate(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+	c := m.(*client.V1Client)
+
+	var diags diag.Diagnostics
+
+	workspace := toWorkspace(d)
+
+	uid, err := c.CreateWorkspace(workspace)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	d.SetId(uid)
+
+	resourceWorkspaceRead(ctx, d, m)
+
+	return diags
+}
+
+func resourceWorkspaceRead(_ context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+	c := m.(*client.V1Client)
+
+	var diags diag.Diagnostics
+
+	uid := d.Id()
+
+	workspace, err := c.GetWorkspace(uid)
+	if err != nil {
+		return diag.FromErr(err)
+	} else if workspace == nil {
+		d.SetId("")
+		return diags
+	}
+
+	fp := flattenWorkspaceClusters(workspace)
+	if err := d.Set("clusters", fp); err != nil {
+		return diag.FromErr(err)
+	}
+
+	if err := d.Set("backup_policy", flattenWorkspaceBackupPolicy(workspace)); err != nil {
+		return diag.FromErr(err)
+	}
+
+	if err := d.Set("cluster_rbac_binding", flattenClusterRBAC(workspace.Spec.ClusterRbacs)); err != nil {
+		return diag.FromErr(err)
+	}
+
+	if err := d.Set("namespaces", flattenWorkspaceClusterNamespaces(workspace.Spec.ClusterNamespaces)); err != nil {
+		return diag.FromErr(err)
+	}
+
+	return diags
+}
+
+func resourceWorkspaceUpdate(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+	c := m.(*client.V1Client)
+
+	var diags diag.Diagnostics
+
+	workspace, err := c.GetWorkspace(d.Id())
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	if d.HasChange("clusters") {
+		// resource allocation should go first because clusters are inside.
+		namespaces := toUpdateWorkspaceNamespaces(d)
+		if err := c.UpdateWorkspaceResourceAllocation(d.Id(), namespaces); err != nil {
+			return diag.FromErr(err)
+		}
+		rbacs := toUpdateWorkspaceRBACs(d)
+		if err := c.UpdateWorkspaceRBACS(d.Id(), workspace.Spec.ClusterRbacs[0].Metadata.UID, rbacs); err != nil {
+			return diag.FromErr(err)
+		}
+
+	} else {
+		rbacs := toUpdateWorkspaceRBACs(d)
+		if d.HasChange("cluster_rbac_binding") {
+			if err := c.UpdateWorkspaceRBACS(d.Id(), workspace.Spec.ClusterRbacs[0].Metadata.UID, rbacs); err != nil {
+				return diag.FromErr(err)
+			}
+		}
+		if d.HasChange("namespaces") {
+			if err := c.UpdateWorkspaceResourceAllocation(d.Id(), toUpdateWorkspaceNamespaces(d)); err != nil {
+				return diag.FromErr(err)
+			}
+		}
+	}
+
+	/*
+		if d.HasChange("backup_policy") {
```

Comment:
@nikchern we will need to check for `d.HasChange("backup_policy")`

Reply:
Implemented.